### PR TITLE
Push logprob generation to LLMEngine

### PR DESCRIPTION
**Pull request by Yard1**
This PR moves the logprob detokenization logic away from the OpenAI server to the LLMEngine, allowing for consistent output between the two. It also is a first step towards making the OpenAI server more lightweight by pushing down some of its responsibilities.

It also ensures the logprob tokens are detokenized with the previous tokens in mind (same as generated tokens), which will make them more accurate.

### CODE REVIEW

File: vllm/sequence.py
Comment (esmeetu) [inline vllm/sequence.py:478]: Is it better to move this to `Logprobs`'s `__eq__`?

**Review by esmeetu — APPROVED** [review]
LGTM!

File: vllm/sequence.py
Comment (Yard1) [reply vllm/sequence.py:478]: good call!

**Comment by AguirreNicolas** [comment]
I pointed out an issue w.r.t to LogProbs in a [previous PR](https://github.com/vllm-project/vllm/pull/2292):

>Please refers to the source of the `ChatCompletionResponseChoice` (`class Choice` [source](https://github.com/openai/openai-python/blob/main/src/openai/types/chat/chat_completion.py#L19)) in OpenAI.
>
>There is is a new class called `ChoiceLogprobs` ([source](https://github.com/openai/openai-python/blob/main/src/openai/types/chat/chat_completion.py#L14)) that is not the same as `LogProbs`.
>
>IMO, first it should be implemented both `ChatCompletionTokenLogprob` and `TopLogprob` ([source](https://github.com/openai/openai-python/blob/main/src/openai/types/chat/chat_completion_token_logprob.py#L26))
>
>API reference about the structure: https://platform.openai.com/docs/api-reference/chat/object

Maybe this PR is a good point to merge/attend this into vLLM ?

**Comment by Yard1** [comment]
@AguirreNicolas IIUC, considering that change would need to be mainly implemented in OpenAI server, I think it should be independent of this PR.

**Comment by Yard1** [comment]
@esmeetu I had to add some extra logic, ptal again

**Review by njhill — COMMENTED** [review]
Thanks @Yard1! I'd realized that something like this was needed while making changes to use a threadpool for tokenization (per https://github.com/vllm-project/vllm/pull/2879#issuecomment-1952990299). I'll wait until this is merged before opening the PR for that.

File: vllm/engine/llm_engine.py
Comment (Yard1) [reply vllm/engine/llm_engine.py:949]: Actually, the OpenAI behavior is exactly that - the logprob token text depends on the previous tokens and is not constant.

File: vllm/engine/arg_utils.py
Comment (esmeetu) [inline vllm/engine/arg_utils.py:33]: Is `max_logprobs` better than this? 
And we can add comments for why default value is 5 (from OpenAI API Reference?).

**Review by esmeetu — APPROVED** [review]
@Yard1 LGTM! Could you merge the latest branch and pass the CI?

File: vllm/engine/llm_engine.py
Comment (njhill) [reply vllm/engine/llm_engine.py:949]: @Yard1 ok, thanks! I'll take a closer look at this.

File: vllm/entrypoints/openai/serving_chat.py
Comment (Yard1) [reply vllm/entrypoints/openai/serving_chat.py:108]: I remember the openai client package was not able to handle errors unless they were the first thing that came out of the endpoint. I think the current version may be more robust, though. Will see if the test can still pass with the previous layout.

I think the slight delay in response is fine

**Comment by njhill** [comment]
@Yard1 I realized that this is an API-breaking change for anyone consuming logprobs via the engine API (it actually broke our integration). I'm not sure what the project stance on this is w.r.t. semantic versioning but at minimum I guess it should be highlighted in the 0.3.4 release notes.

**Comment by pcmoritz** [comment]
@Yard1 This also breaks https://github.com/EleutherAI/lm-evaluation-harness -- we should either fix the harness or roll back the API change :)

```
  File "/home/ray/anaconda3/bin/lm_eval", line 8, in <module>
    sys.exit(cli_evaluate())
  File "/home/ray/default/lm-evaluation-harness/lm_eval/__main__.py", line 318, in cli_evaluate
    results = evaluator.simple_evaluate(
  File "/home/ray/default/lm-evaluation-harness/lm_eval/utils.py", line 288, in _wrapper
    return fn(*args, **kwargs)
  File "/home/ray/default/lm-evaluation-harness/lm_eval/evaluator.py", line 230, in simple_evaluate
    results = evaluate(
  File "/home/ray/default/lm-evaluation-harness/lm_eval/utils.py", line 288, in _wrapper
    return fn(*args, **kwargs)
  File "/home/ray/default/lm-evaluation-harness/lm_eval/evaluator.py", line 368, in evaluate
    resps = getattr(lm, reqtype)(cloned_reqs)
  File "/home/ray/default/lm-evaluation-harness/lm_eval/api/model.py", line 321, in loglikelihood
    return self._loglikelihood_tokens(new_reqs)
  File "/home/ray/default/lm-evaluation-harness/lm_eval/models/vllm_causallms.py", line 379, in _loglikelihood_tokens
    answer = self._parse_logprobs(
  File "/home/ray/default/lm-evaluation-harness/lm_eval/models/vllm_causallms.py", line 416, in _parse_logprobs
    continuation_logprobs = sum(
TypeError: unsupported operand type(s) for +: 'int' and 'Logprob'
```

**Comment by Yard1** [comment]
We should fix the harness.

**Comment by pcmoritz** [comment]
Sounds good to me -- can you make a PR for it?

**Pull request merged**
